File: packages/runtime-python/src/copilotkit_runtime/telemetry.py
"""Canonical CopilotKit analytics with bounded asynchronous delivery."""

import asyncio
import base64
import inspect
import json
import math
import os
import random
import re
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, NoReturn
from urllib.parse import urlsplit

import httpx

from .models import Json

EventSink = Callable[[Json], Awaitable[None]]
_PREFIX = "oss.runtime."
_IDENTITY = re.compile(r"^[A-Za-z0-9_-]{1,128}$")
# ECMAScript WhiteSpace + LineTerminator; Python str.strip() differs for FEFF/NEL.
_JS_WHITESPACE = "\u0009\u000a\u000b\u000c\u000d\u0020\u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000\ufeff"


def _invalid_json_constant(value: str) -> NoReturn:
    """Reject non-JSON numeric constants that Python's decoder otherwise accepts."""
    raise ValueError("Invalid JSON constant")


def _license_identity(token: str | None) -> str | None:
    """Read a safe legacy analytics claim, never verify a license or retain its token."""
    if not isinstance(token, str):
        return None
    parts = token.split(".")
    if len(parts) != 3:
        return None
    payload = parts[1]
    if not re.fullmatch(r"[A-Za-z0-9_-]+", payload) or len(payload) % 4 == 1:
        return None
    try:
        decoded = json.loads(
            base64.urlsafe_b64decode(payload + "=" * (-len(payload) % 4)).decode(
                "utf-8", errors="replace"
            ),
            parse_constant=_invalid_json_constant,
        )
        value = decoded.get("telemetry_id") if isinstance(decoded, dict) else None
        if isinstance(value, str) and _IDENTITY.fullmatch(value.strip(" \t")):
            return value.strip(" \t")
    except (ValueError, TypeError, RecursionError):
        pass
    return None


@dataclass(frozen=True)
class TelemetryStats:
    """Local exporter diagnostics; these counters do not create analytics traffic."""

    queued: int
    sent: int
    failed: int
    dropped: int
    sampled_out: int


class Telemetry:
    """Sample canonical events and enqueue them without delaying request handling."""

    def __init__(
        self,
        enabled: bool = True,
        sink: EventSink | None = None,
        *,
        sample_rate: float = 0.05,
        telemetry_id: str | None = None,
        license_token: str | None = None,
        url: str = "https://telemetry.copilotkit.ai/ingest",
        queue_capacity: int = 256,
        timeout: float = 3,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self.enabled = enabled and not any(
            os.getenv(key, "").lower() in ("true", "1")
            for key in ("DO_NOT_TRACK", "COPILOTKIT_TELEMETRY_DISABLED")
        )
        if sink is not None and not (
            inspect.iscoroutinefunction(sink)
            or inspect.iscoroutinefunction(getattr(sink, "__call__", None))
        ):
            raise ValueError("Telemetry sink must be async")
        override = os.getenv("COPILOTKIT_TELEMETRY_SAMPLE_RATE")
        self.sample_rate = float(override) if override else sample_rate
        if not math.isfinite(self.sample_rate) or not 0 <= self.sample_rate <= 1:
            raise ValueError("Sample rate must be finite and between 0 and 1")
        if queue_capacity < 1 or not math.isfinite(timeout) or not 0 < timeout <= 3:
            raise ValueError("Telemetry queue must be positive and timeout within (0, 3] seconds")
        self.telemetry_id = next(
            (
                candidate.strip(" \t")
                for candidate in (telemetry_id, os.getenv("CPK_TELEMETRY_ID"))
                if isinstance(candidate, str) and _IDENTITY.fullmatch(candidate.strip(" \t"))
            ),
            None,
        )
        self.identified = False
        if self.telemetry_id is None:
            self.telemetry_id = _license_identity(
                next(
                    (
                        candidate
                        for candidate in (license_token, os.getenv("COPILOTKIT_LICENSE_TOKEN"))
                        if isinstance(candidate, str) and candidate.strip(_JS_WHITESPACE)
                    ),
                    None,
                )
            )
            self.identified = self.telemetry_id is not None
            if self.identified:
                self.sample_rate = 1
        self.url = os.getenv("COPILOTKIT_TELEMETRY_URL") or url
        parsed = urlsplit(self.url)
        if parsed.scheme not in ("http", "https") or not parsed.hostname or parsed.username:
            raise ValueError("Invalid telemetry endpoint")
        self.sink = sink
        self.timeout = timeout
        self._queue: asyncio.Queue[Json] = asyncio.Queue(queue_capacity)
        self._worker: asyncio.Task[None] | None = None
        self._client = http_client
        self._owns_client = http_client is None
        self._closed = False
        self._sent = self._failed = self._dropped = self._sampled_out = 0

    @property
    def stats(self) -> TelemetryStats:
        """Return a snapshot of queue pressure and transport outcomes."""
        return TelemetryStats(
            self._queue.qsize(), self._sent, self._failed, self._dropped, self._sampled_out
        )

    async def emit(self, name: str, **attributes: Any) -> None:
        """Accept only canonical events and construct a fixed, content-free payload."""
        if not self.enabled or self._closed:
            return
        event = name.removeprefix(_PREFIX)
        properties: Json
        if event == "instance_created":
            count = attributes.get("agentsAmount", 0)
            properties = {
                "actionsAmount": 0,
                "endpointTypes": [],
                "endpointsAmount": 0,
                "agentsAmount": count if type(count) is int and count >= 0 else 0,
                "cloud.api_key_provided": False,
            }
        elif event == "copilot_request_created" and attributes.get("requestType") in (
            "run",
            "connect",
        ):
            properties = {
                "requestType": attributes["requestType"],
                "cloud.guardrails.enabled": False,
                "cloud.api_key_provided": False,
            }
        elif event in ("agent_execution_stream_started", "agent_execution_stream_ended"):
            properties = {}
        elif event == "agent_execution_stream_errored":
            properties = {
                "error": attributes.get("error")
                if attributes.get("error")
                in ("AGENT_EXECUTION_FAILED", "RUN_STOPPED", "GATEWAY_START_FAILED")
                else "AGENT_EXECUTION_FAILED"
            }
        else:
            return
        if self.sample_rate == 0 or random.random() >= self.sample_rate:
            self._sampled_out += 1
            return
        envelope = {
            "event": _PREFIX + event,
            "properties": properties,
            "ts": int(time.time()),
            "package": {"name": "copilotkit-intelligence-runtime", "version": "0.1.0"},
            "global_properties": {
                "sampleRate": self.sample_rate,
                "sampleRateAdjustmentFactor": 1 - self.sample_rate,
                "sampleWeight": 1 / self.sample_rate,
                "telemetry_identified": self.identified,
                "telemetry_emitter": "native-python",
                "telemetry_transport": "lambda",
            },
        }
        try:
            self._queue.put_nowait(envelope)
        except asyncio.QueueFull:
            self._dropped += 1
            return
        if self._worker is None or self._worker.done():
            self._worker = asyncio.create_task(self._drain(), name="copilotkit-telemetry")

    async def _drain(self) -> None:
        """Send one item at a time with bounded timeout and no automatic redirects."""
        while True:
            envelope = await self._queue.get()
            try:
                async with asyncio.timeout(self.timeout):
                    if self.sink:
                        await self.sink(envelope)
                    else:
                        if self._client is None:
                            self._client = httpx.AsyncClient(follow_redirects=False)
                        headers = {
                            "Content-Type": "application/json",
                            "User-Agent": "CopilotKit-Runtime/0.1.0 (copilotkit-intelligence-runtime)",
                        }
                        if self.telemetry_id:
                            headers["X-CopilotKit-Telemetry-Id"] = self.telemetry_id
                        response = await self._client.post(
                            self.url,
                            json=envelope,
                            headers=headers,
                            timeout=self.timeout,
                            follow_redirects=False,
                        )
                        response.raise_for_status()
                self._sent += 1
            except Exception:
                self._failed += 1
            finally:
                self._queue.task_done()

    async def flush(self, timeout: float = 3) -> bool:
        """Wait at most timeout seconds for the existing queue; return whether it drained."""
        try:
            async with asyncio.timeout(timeout):
                await self._queue.join()
            return True
        except TimeoutError:
            return False

    async def aclose(self, timeout: float = 3) -> None:
        """Stop accepting events and discard queued work after the shutdown deadline."""
        if self._closed:
            return
        self._closed = True
        await self.flush(timeout)
        if self._worker:
            self._worker.cancel()
            await asyncio.gather(self._worker, return_exceptions=True)
        while not self._queue.empty():
            self._queue.get_nowait()
            self._queue.task_done()
            self._dropped += 1
        if self._owns_client and self._client:
            await self._client.aclose()

File: packages/runtime-ruby/lib/copilotkit/telemetry.rb
# frozen_string_literal: true
require 'timeout'
require 'base64'

module CopilotKit
  # Bounded asynchronous analytics exporter with TypeScript-compatible envelopes.
  # Every attribute is constructed here; application content is never copied.
  class Telemetry
    PREFIX = 'oss.runtime.'
    ENDPOINT = 'https://telemetry.copilotkit.ai/ingest'
    EVENTS = %w[instance_created copilot_request_created agent_execution_stream_started agent_execution_stream_ended agent_execution_stream_errored].freeze

    def initialize(exporter: nil, disabled: false, sample_rate: 0.05, telemetry_id: nil, license_token: nil,
                   url: nil, queue_capacity: 256, random: -> { Random.rand }, env: ENV)
      @disabled = disabled || %w[DO_NOT_TRACK COPILOTKIT_TELEMETRY_DISABLED].any? { |key| %w[true 1].include?(env[key].to_s.downcase) }
      configured_rate = env.key?('COPILOTKIT_TELEMETRY_SAMPLE_RATE') ? env['COPILOTKIT_TELEMETRY_SAMPLE_RATE'] : sample_rate
      begin
        @rate = Float(configured_rate)
      rescue ArgumentError, TypeError
        @rate = 0.05
      end
      @rate = 0.05 unless @rate.finite? && @rate.between?(0, 1)
      @id = [telemetry_id, env['CPK_TELEMETRY_ID']].filter_map do |value|
        next unless value.is_a?(String)
        normalized = value.gsub(/\A[ \t]+|[ \t]+\z/, '')
        normalized if normalized.match?(/\A[A-Za-z0-9_-]{1,128}\z/)
      end.first unless @disabled
      @identified = false
      unless @disabled || @id
        token = [license_token, env['COPILOTKIT_LICENSE_TOKEN']].find do |value|
          value.is_a?(String) && value.match?(/[^\u0009-\u000D\u0020\u00A0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]/)
        end
        @id = license_telemetry_id(token)
        @identified = !@id.nil?
        @rate = 1.0 if @identified
      end
      @url = url || env['COPILOTKIT_TELEMETRY_URL'] || ENDPOINT
      @exporter, @random = exporter, random
      raise ArgumentError, 'queue_capacity must be positive' unless queue_capacity.is_a?(Integer) && queue_capacity.positive?
      @queue, @mutex, @condition = SizedQueue.new(queue_capacity), Mutex.new, ConditionVariable.new
      @pending, @closed, @worker = 0, false, nil
    end

    def disabled?
      @disabled
    end

    # Queue one sampled event without waiting for the network. A full queue drops it.
    def emit(name, attributes = {})
      return if @disabled || (!@identified && (@rate.zero? || @random.call >= @rate))
      event_name = name.delete_prefix(PREFIX)
      return unless name.start_with?(PREFIX) && EVENTS.include?(event_name)
      properties = case event_name
                   when 'instance_created'
                     count = attributes['agentsAmount']
                     { 'actionsAmount' => 0, 'endpointTypes' => [], 'endpointsAmount' => 0,
                       'agentsAmount' => count.is_a?(Integer) && count >= 0 ? count : 0, 'cloud.api_key_provided' => false }
                   when 'copilot_request_created'
                     return unless %w[run connect].include?(attributes['requestType'])
                     { 'requestType' => attributes['requestType'], 'cloud.guardrails.enabled' => false, 'cloud.api_key_provided' => false }
                   when 'agent_execution_stream_errored'
                     { 'error' => 'AGENT_RUN_FAILED' }
                   else
                     {}
                   end
      event = { 'event' => name, 'properties' => properties, 'ts' => Time.now.to_i,
                'package' => { 'name' => 'copilotkit-runtime-ruby', 'version' => '0.1.0' },
                'global_properties' => { 'sampleRate' => @rate, 'sampleRateAdjustmentFactor' => 1 - @rate,
                  'sampleWeight' => 1 / @rate, 'telemetry_identified' => @identified,
                  'telemetry_emitter' => 'runtime-ruby', 'telemetry_transport' => 'lambda' } }
      @mutex.synchronize do
        return if @closed
        begin
          @queue.push(event, true)
        rescue ThreadError
          return
        end
        @pending += 1
        @worker ||= Thread.new { consume }
      end
      nil
    rescue StandardError
      nil
    end

    # Wait for already queued events, up to the caller's deadline.
    def flush(timeout: 3)
      deadline = monotonic + timeout
      @mutex.synchronize do
        while @pending.positive?
          remaining = deadline - monotonic
          return false unless remaining.positive?
          @condition.wait(@mutex, remaining)
        end
      end
      true
    end

    # Stop accepting work, drain within a deadline, then cancel a stalled exporter.
    def close(timeout: 3)
      worker = @mutex.synchronize do
        return if @closed
        @closed = true
        begin
          @queue.push(nil, true) if @worker
        rescue ThreadError
          # A full queue drains naturally; the worker stops when closed and empty.
        end
        @worker
      end
      worker&.join(timeout)
      if worker&.alive?
        worker.kill
        worker.join(0.1)
      end
      nil
    rescue StandardError
      nil
    end

    private

    # Claims provide analytics attribution only, never license verification.
    def license_telemetry_id(token)
      return unless token.is_a?(String)
      parts = token.split('.', -1)
      return unless parts.length == 3
      payload = parts[1]
      return unless payload.match?(/\A[A-Za-z0-9_-]+\z/) && payload.length % 4 != 1
      decoded = JSON.parse(Base64.urlsafe_decode64(payload))
      return unless decoded.is_a?(Hash) && decoded['telemetry_id'].is_a?(String)
      id = decoded['telemetry_id'].gsub(/\A[ \t]+|[ \t]+\z/, '')
      id if id.match?(/\A[A-Za-z0-9_-]{1,128}\z/)
    rescue ArgumentError, JSON::ParserError
      nil
    end

    def monotonic
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end

    def consume
      loop do
        event = @queue.pop
        break unless event
        begin
          Timeout.timeout(3) { @exporter ? @exporter.call(event) : send_http(event) }
        rescue StandardError
          # Analytics failure never changes runtime behavior or emits raw diagnostics.
        ensure
          @mutex.synchronize { @pending -= 1; @condition.broadcast }
        end
        break if @mutex.synchronize { @closed && @queue.empty? }
      end
    ensure
      begin
        Timeout.timeout(3) { @exporter.close } if @exporter.respond_to?(:close)
      rescue StandardError
        nil
      end
    end

    def send_http(event)
      uri = URI(@url)
      return unless uri.is_a?(URI::HTTP) && !uri.userinfo
      headers = { 'content-type' => 'application/json' }
      headers['X-CopilotKit-Telemetry-Id'] = @id if @id
      request = Net::HTTP::Post.new(uri.request_uri, headers)
      request.body = JSON.generate(event)
      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https', open_timeout: 3, read_timeout: 3) do |http|
        http.max_retries = 0
        http.request(request) do |response| # Net::HTTP never follows redirects.
          received = 0
          response.read_body do |chunk|
            received += chunk.bytesize
            raise IOError, 'Telemetry response exceeded size limit' if received > 65_536
          end
        end
      end
    end
  end
end
